keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Paths
+
+The overarching `/openapi/v3` endpoint will contain the list of paths (groups) available and serve as a discovery endpoint. Clients can then choose the group(s) to fetch and send additional requests.
+
+/openapi/v3
+
+```json
+{
+   "Paths" : [
+      "api",
+      "api/v1",
+      "apis",
+      "apis/admissionregistration.k8s.io",
+      "apis/apiextensions.k8s.io",
+      "apis/apps",
+      "apis/authentication.k8s.io",
+      ...
+   ]
+}
+```
+
+Based on the provided group, clients can then request `openapi/v3/apis/apps`, `/openapi/v3/api/networking.k8s.io` and etc. These leaf node specs are self contained OpenAPI v3 specs and include all referenced types.
+
+### Controllers
+
+#### OpenAPI Builder
+
+The OpenAPI builder for CRDs converts the v3 validation schema into a v2 schema. We will keep the conversion to the v2 schema for backwards compatibility, but also generate the v3 schema to publish at the new v3 endpoint.
+
+#### Proto Models & ETags
+
+Kubernetes publishes the openapi schema in both JSON and Protobuf. v3 schemas will include the same mechanisms as v2, publishing a protobuf model of the schema, as well as the corresponding etags for caching. In order to publish the schema based on groups, the CRDs will be grouped by group and published only at the endpoint for their specific group.
+
+#### Aggregator
+
+The aggregator has a mapping of all the APIServices and refreshes the aggregated spec if a single spec or CRD changes. Since the underlying api services will publish resources based on group, the aggregator must be able to discover all group endpoints for each APIService. This will increase the number of requests on bootstrap and subsequently the number of internal requests on a schema change. However, since aggregation is done by group, only a small subset of groups will be invalidated and reaggregated.
+
+### OpenAPI
+
+Swagger 2.0 and OpenAPI 3.0 have a couple of incompatible changes that are not currently supported by kube-openapi. Definitions are moved to components and the schema for Paths is slightly modified. A new struct and builder will be created in kube-openapi to represent the v3 schema to be imported by the necessary consumers.
+
+CRD structural schemas will be published as is without any value stripping for v3. Built-in types will generate the v3 OpenAPI definition, and either keep all fields or strip incompatible fields when building the swagger spec for v3 and v2 respectively.
+
+The OpenAPI handler will have an additional map to handle the publishing of v3.
+
+```go
+// Group -> OpenAPIv3 mapping
+var v3Schemas map[string]*OpenAPIv3
+
+type OpenAPIv3 struct {
+  rwMutex sync.RWMutex
+  lastModified time.Time
+
+  spec3Bytes []byte
+  spec3Pb []byte
+  spec3PbGz []byte
+
+  spec3BytesETag string
+  spec3PbETag string
+  spec3PbGzETag string
+}
+```
+
```

Comment:
I miss a plan about our native resources: in v2 we publish invalid specs today, e.g. quantities are not specified via anyOf afaik, and maybe other polymorphic types are similar. Also afaik we are not correct with nullable at places in ObjectMeta when I remember right. We should use this chance to fix those issues.